### PR TITLE
Improve XPathPathExpr code quality

### PR DESCRIPTION
**Pull request by dcbriccetti**
Move eval method from XPathPathExpr to new class, XPathPathExprEval, and modernize its code and make it much more understandable.

(cherry picked from commit 2d5998d)

#### What has been done to verify that this works as intended?
I saw that all tests still pass, and I stepped through the code in the debugger.

#### Why is this the best possible solution? Were any other approaches considered?
The goal was to separate the `eval` concern from the rest of XPathPathExpr, and to modernize code using automatic IDEA refactorings.

#### Are there any risks to merging this code? If so, what are they?
The risk seems low. I used refactorings like extract method, which are safe.

### CODE REVIEW

File: src/org/javarosa/xpath/expr/XPathPathExpr.java [hunk 1]
```diff
@@ -70,209 +69,134 @@
...
-        //to fix conditions based on non-relevant data, filter the nodeset by relevancy
-        for (int i = 0; i < nodesetRefs.size(); i++) {
-            if (!m.resolveReference(nodesetRefs.get(i)).isRelevant()) {
-                nodesetRefs.remove(i);
-                i--;
-            }
-        }
-
-        return new XPathNodeset(nodesetRefs, m, ec);
+    public XPathNodeset eval(final DataInstance unusedDataInstance, final EvaluationContext ec) {
```
Comment (ggalmazor) [inline src/org/javarosa/xpath/expr/XPathPathExpr.java]: I would argue to remove these `final`s here. Maybe the one for `TreeReference ref` in `getReference()` is useful while it gets smaller but here it's just one line of code and those variables are just passed downstream.

Comment (dcbriccetti) [reply src/org/javarosa/xpath/expr/XPathPathExpr.java]: Thanks. I intended to remove them, and will now. They were part of the evolution of the code (originally the first parameter was never read, but reassigned!).

File: src/org/javarosa/xpath/expr/XPathPathExpr.java [hunk 2]
```diff
@@ -355,16 +285,16 @@ public String toString () {
...
+            XPathPathExpr x = (XPathPathExpr) o;
 
             //Shortcuts for easily comparable values
-            if(init_context != x.init_context || steps.length != x.steps.length) {
+            if (init_context != x.init_context || steps.length != x.steps.length) {
                 return false;
             }
 
-            return ExtUtil.arrayEquals(steps, x.steps) && (init_context == INIT_CONTEXT_EXPR ? filtExpr.equals(x.filtExpr) : true);
+            return ExtUtil.arrayEquals(steps, x.steps) && (init_context != INIT_CONTEXT_EXPR || filtExpr.equals(x.filtExpr));
```
Comment (ggalmazor) [inline src/org/javarosa/xpath/expr/XPathPathExpr.java]: I might have to wait for tomorrow to be fresh again to really be sure that this does the same, but I'm guessing you've just let IntelliJ simplify the boolean expression :+1:

Comment (dcbriccetti) [reply src/org/javarosa/xpath/expr/XPathPathExpr.java]: Yes.

File: src/org/javarosa/xpath/expr/XPathPathExpr.java [hunk 3]
```diff
@@ -70,209 +69,134 @@
...
+            case XPathPathExpr.INIT_CONTEXT_EXPR:
+                if (filtExpr.x instanceof XPathFuncExpr) {
+                    XPathFuncExpr func = (XPathFuncExpr) (this.filtExpr.x);
+                    switch (func.id.toString()) {
+                        case "instance":
+                            ref.setRefLevel(TreeReference.REF_ABSOLUTE); //i assume when refering the non main instance you have to be absolute
+
+                            parentsAllowed = false;
+                            if (func.args.length != 1) {
+                                throw new XPathUnsupportedException("instance() function used with " + func.args.length + " arguments. Expecting 1 argument");
```
Comment (ggalmazor) [inline src/org/javarosa/xpath/expr/XPathPathExpr.java]: "arguement" sounds much more interesting. Spelling is overrated :D

File: src/org/javarosa/xpath/expr/XPathPathExpr.java (hunk 1 above)
Comment (ggalmazor) [reply src/org/javarosa/xpath/expr/XPathPathExpr.java]: Wow, it's great that you fixed that one :)

Comment (dcbriccetti) [reply src/org/javarosa/xpath/expr/XPathPathExpr.java]: At least I made it obvious that all those callers are passing in something that is ignored.

**Pull request merged**